### PR TITLE
nixos/light: add control options

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

As discussed in https://github.com/NixOS/nixpkgs/issues/57602 and https://github.com/NixOS/nixpkgs/pull/58609

cc @primeos @oxij in case you're interested

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/programs/light.nix [hunk 1]
```diff
@@ -17,11 +18,82 @@ in
...
+        { keys = [ 225 ]; events = [ "key" "rep" ]; command = "${light} -A ${toString cfg.mediaKeys.step}"; }
+      ];
+    };
+
+    system.activationScripts.setup-light = ''
+      ${light} -N ${toString cfg.minimumBrightness}
+      ${light} -O
+    '';
+
+    systemd.services.light = {
```
Comment (grahamc) [inline nixos/modules/programs/light.nix]: This should not be necessary, as systemd already handles this.

File: nixos/modules/programs/light.nix [hunk 2]
```diff
@@ -17,11 +18,82 @@ in
...
   };
 
   config = mkIf cfg.enable {
     environment.systemPackages = [ pkgs.light ];
     services.udev.packages = [ pkgs.light ];
+
+    services.actkbd = mkIf cfg.mediaKeys.enable {
+      enable = true;
+      bindings = [
+        { keys = [ 224 ]; events = [ "key" "rep" ]; command = "${light} -U ${toString cfg.mediaKeys.step}"; }
```
Comment (grahamc) [inline nixos/modules/programs/light.nix]: I'm not sure we actually want to include these sorts of options in NixOS as they increase evaluation time for each user, and are not very complicated. I would prefer we add documentation to the wiki on how to do this.

File: nixos/modules/programs/light.nix [hunk 3]
```diff
@@ -4,6 +4,7 @@ with lib;
 
 let
   cfg = config.programs.light;
+  light = if cfg.devicePath != null then "${pkgs.light}/bin/light -s ${cfg.devicePath}" else "${pkgs.light}/bin/light";
```
Comment (Synthetica9) [inline nixos/modules/programs/light.nix]: I'd do something like the following (untested):

```nix
light = "${pkgs.light}/bin/light" + lib.optionalString (cfg.devicePath != null) " -s ${cfg.devicePath}";
```

File: nixos/modules/programs/light.nix (hunk 1 above)
Comment (grahamc) [reply nixos/modules/programs/light.nix]: https://www.freedesktop.org/software/systemd/man/systemd-backlight@.service.html

File: nixos/modules/programs/light.nix (hunk 2 above)
Comment (Infinisil) [reply nixos/modules/programs/light.nix]: Or maybe this is better fit for [NUR](https://github.com/nix-community/NUR)

Comment (gnidorah) [reply nixos/modules/programs/light.nix]: @grahamc What's the purpose of light module then? Also there is an inconsistency for laptop users. We have `sound.mediaKeys` but have no brightness keys control.

File: nixos/modules/programs/light.nix (hunk 1 above)
Comment (gnidorah) [reply nixos/modules/programs/light.nix]: @grahamc Thanks!

File: nixos/modules/programs/light.nix (hunk 3 above)
Comment (gnidorah) [reply nixos/modules/programs/light.nix]: @Synthetica9 Applied, thanks!

**Comment by oxij** [comment]
> This should not be necessary, as systemd already handles this.

This has the advantage of not needing systemd, being configurable (instead of using some magic clamping), and setting brightness even before systemd starts, which seems like a nice addition to me.

Though, the problem in NixOS is that with systemd, systemd will override the backlight anyway, so I agree that either this piece of code does not belong to NixOS, or, alternatively, this should disable `systemd-backlight` when enabled.

> I'm not sure we actually want to include these sorts of options in NixOS as they increase evaluation time for each user, and are not very complicated. I would prefer we add documentation to the wiki on how to do this.

By that logic NixOS should not have any optional services in `baseModules` (which was discussed multiple times before and rejected). They might not be very complicated, but I imagine that enabling one declarative option as opposed to copy-pasting the same code between configs is the reason NixOS exists.

My only other nitpick is the same as the one by @Synthetica9 above.

Otherwise, this LGTM.

**Comment by gnidorah** [comment]
@oxij 

> This has the advantage of not needing systemd, being configurable (instead of using some magic clamping), and setting brightness even before systemd starts, which seems like a nice addition to me.

Right. You can disable clamping with `ENV{ID_BACKLIGHT_CLAMP}="0"` in udev rules. But why to bother, if light is able to save and restore brightness level by itself?

> Though, the problem in NixOS is that with systemd, systemd will override the backlight anyway, so I agree that either this piece of code does not belong to NixOS, or, alternatively, this should disable `systemd-backlight` when enabled.

I would disable `systemd-backlight`, just yet not found a way to do this.

**Comment by grahamc** [comment]
Overall, these changes seem more suited for either documentation or NUR as @Infinisil suggested.

**Comment by grahamc** [comment]
The saving / restoring GC code doesn't belong, as systemd handles that for us.

The device path seem reasonable in the context of keybindings, but keybindings enshrined in the module also seems off-topic for NixOS's services.

Defining keybindings in the core of NixOS is essentially building a desktop environment in our services. Our services should be lighter and more generic. These settings should be in documentation, examples, or NUR.

**Comment by oxij** [comment]
@grahamc

> The saving / restoring GC code doesn't belong, as systemd handles that for us.

I don't see how that applies to this module. systemd has timesync or whatever, but we have ntp and chrony modules, systemd has resolved or whatever, but we still have dhcpcd and dnsmasq, it has networkd but we have scripted networking.

This provides an alternative to systemd backlight without non-configurable defaults given by the God himself, which is the whole point.

> The device path seem reasonable in the context of keybindings, but keybindings enshrined in the module also seems off-topic for NixOS's services.
>
> Defining keybindings in the core of NixOS is essentially building a desktop environment in our services. Our services should be lighter and more generic. These settings should be in documentation, examples, or NUR.

By that logic you should remove a lot of NixOS modules, e.g, from the top of my head ./nixos/modules/services/web-servers/lighttpd/cgit.nix enshrines cgit URLs, most modules enshrine common /var and /run paths, we enshrine most uids and usernames, etc etc.

I don't see any technical arguments against enshrining good defaults if the actual value does not really matter or you can configure and/or disable them easily, which, btw, this module does and systemd builtin does not.

@gnidorah

I plan to adopt this to SLNOS notchb branch anyways, so do not get discouraged. If you finish it, I'll take it (and if you don't, I'll take it and finish it myself later).

**Comment by gnidorah** [comment]
> If you finish it, I'll take it (and if you don't, I'll take it and finish it myself later).

Please let me know what should I fix there.

**Comment by oxij** [comment]
> Please let me know what should I fix there.

I still think systemd unit needs to be disabled. But it seems this is not trivial and requires some hacking in nixos/modules/system/boot/systemd.nix (which I already did and will PR soon).

**Comment by oxij** [comment]
Systemd module change: #59827.

The polish of this built on top of the systemd change:

https://github.com/oxij/nixpkgs/commits/pull/59742-light-backlight
https://github.com/oxij/nixpkgs/commit/54cfd0b21c758c0ca5a0d03fd17a85be78441f11

IMHO the complexity of it warrants it to be merged to NixOS despite whatever anyone else says, making it work properly without conflicts and race conditions was quite a pain in the ass because everyone and their grandmother wrestles for backlight control on Linux+systemd. Good luck describing that in a wiki page without using exactly the code from my branch.

The end result in awesome, though. Finally, a proper backlight control via intel_backlight (i.e. properly granular, not just 15 ACPI steps) everywhere, including the linux ttys.

**Pull request closed without merging**